Added authentication to POST endpoint

morses-code opened the pull request:
### What
Added authentication to POST endpoint

### How to review
Check tests pass and review implementation of authentication.

Run publishing services and dp-sessions-api (make debug) and test the following curl command (replace service auth token, with your valid one, also test without auth header and with an invalid header):
- curl --location --request POST 'localhost:24400/session' --header 'Authorization: ${SERVICE_AUTH_TOKEN}' --header 'Content-Type: application/json' --data-raw '{
	"email":"test@test.com"
}'

morses-code committed:
Starting adding auth to sessions api

- Added AuthHandler interface
- Updated Setup parameters
- Added auth to POST method
- Updated api_test to use auth
- Created AuthHandler mock
- Added Zebedee URL to config

morses-code committed:
Auth added to POST endpoint

- Created AuthHandler interface
- Wrapped POST endpoint with auth 
- Added auth in api_test
- Created Auth mock
- Added Zebedee and ServiceAuthToken to config

morses-code committed:
Merge remote-tracking branch 'origin/feature/sessions-auth' into feature/sessions-auth

morses-code committed:
Correcting imports

jessjenkins commented on api/api.go:
```diff
@@ -1,7 +1,12 @@
 package api
 
+//go:generate moq -out mockauth_test.go . AuthHandler
```
Can I suggest generating the mock into a `mock` subdir? (eg. https://github.com/ONSdigital/dp-image-api/tree/develop/service/mock) I like the idea of generated sources being kept separate from authored sources.

morses-code replied:
Moved all the mocks into their own sub directory, moved all interfaces from api.go and handlers.go to interfaces.go. Changed the _test packages to api_test to avoid the circular imports. (looks a lot better structured like this)

jessjenkins commented on service/service.go:
```diff
@@ -2,7 +2,10 @@ package service
+	"github.com/ONSdigital/dp-api-clients-go/zebedee"
+	"github.com/ONSdigital/dp-authorisation/auth"
 	"github.com/ONSdigital/dp-healthcheck/healthcheck"
+	rchttp "github.com/ONSdigital/dp-rchttp"
```
`rchttp` is deprecated. Consider `dp-net` instead.

morses-code replied:
Updated to dp-net/http

justinpjose commented on api/api.go:
```diff
@@ -11,15 +16,20 @@ type API struct {
+	create := auth.Permissions{Create: true}
 
-	r.HandleFunc("/session", CreateSessionHandlerFunc(nopSess, nopCache)).Methods("POST")
+	r.HandleFunc("/session", permissions.Require(create, CreateSessionHandlerFunc(nopSess, nopCache))).Methods("POST")
```
The naming of the endpoint should be plural so `/sessions` as we will be dealing with multiple sessions ideally and it is the conventional way of naming it

morses-code replied:
You are right, standard does suggest to pluralise the endpoints so have changed it.

justinpjose commented:
So all your tests passes which is good news.
Also, tested your API locally with the following conditions and they all worked as expected:
- POST request with No Authentication Header: 400 Bad Request [PASSED]
- POST request with Invalid Authentication Header: 401 Unauthorized [PASSED]
- POST request with Valid Authentication Header: 201 Created [PASSED]
- GET request with No Authentication Header: 200 OK [PASSED]

Just need to do changes mentioned by myself and Jess and all good to go :)

daiLlew commented on config/config.go:
```diff
@@ -11,6 +11,8 @@ type Config struct {
 	HealthCheckInterval        time.Duration `envconfig:"HEALTHCHECK_INTERVAL"`
 	HealthCheckCriticalTimeout time.Duration `envconfig:"HEALTHCHECK_CRITICAL_TIMEOUT"`
+	ZebedeeURL                 string        `envconfig:"ZEBEDEE_URL"`
+	ServiceAuthToken           string        `envconfig:"SERVICE_AUTH_TOKEN"`
```
There should be an omit json tag on this field as its sensitive information.

morses-code replied:
ServiceAuthToken should now be omitted.

daiLlew commented on service/service.go:
```diff
@@ -86,7 +95,32 @@ func (svc *Service) Close(ctx context.Context) {
+}
+
+func getAuthorisationHandlers(cfg *config.Config) api.AuthHandler {
+	auth.LoggerNamespace("dp-sessions-api-auth")
```
I think this is generally the same as the app name.

morses-code replied:
I copied this from the observation api - https://github.com/ONSdigital/dp-observation-api/blob/develop/service/service.go, it's also like this in the example in - https://github.com/ONSdigital/dp-authorisation/blob/master/example/main.go

morses-code committed:
Moved mocks to own subdir and pluralised endpoints

- Moved AuthHandler interface from api.go > interfaces.go
- Pluralised endpoints
- *_test.go files now are in the api_test package
- Updated tests to import package changes
- Moved Sessions and Cache interfaces from handlers.go > interfaces.go
- Created interfaces.go file
- Moved mocks to own package and subdir

morses-code committed:
Update deprecated import

- Using dp-net/http instead of dp-rchttp

morses-code committed:
Omitting fields from json

- omitted ZebedeeURL and ServiceAuthToken from json

morses-code committed:
omitted one to many fields
